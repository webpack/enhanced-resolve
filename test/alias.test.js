const path = require("path");
const { Volume } = require("memfs");
const { ResolverFactory } = require("../");
const CachedInputFileSystem = require("../lib/CachedInputFileSystem");
const fs = require("fs");
const {
	posixSep,
	absoluteOsBasedResolvedPath,
	absoluteOsBasedPath,
	obps
} = require("./util/path-separator");

const nodeFileSystem = new CachedInputFileSystem(fs, 4000);

describe("alias", () => {
	let resolver;

	beforeEach(() => {
		const fileSystem = Volume.fromJSON(
			{
				[`${absoluteOsBasedPath}a${obps}index`]: "",
				[`${absoluteOsBasedPath}a${obps}dir${obps}index`]: "",
				[`${absoluteOsBasedPath}recursive${obps}index`]: "",
				[`${absoluteOsBasedPath}recursive${obps}dir${obps}index`]: "",
				[`${absoluteOsBasedPath}b${obps}index`]: "",
				[`${absoluteOsBasedPath}b${obps}dir${obps}index`]: "",
				[`${absoluteOsBasedPath}c${obps}index`]: "",
				[`${absoluteOsBasedPath}c${obps}dir${obps}index`]: "",
				[`${absoluteOsBasedPath}d${obps}index.js`]: "",
				[`${absoluteOsBasedPath}d${obps}dir${obps}.empty`]: "",
				[`${absoluteOsBasedPath}e${obps}index`]: "",
				[`${absoluteOsBasedPath}e${obps}anotherDir${obps}index`]: "",
				[`${absoluteOsBasedPath}e${obps}dir${obps}file`]: ""
			},
			absoluteOsBasedPath
		);
		resolver = ResolverFactory.createResolver({
			alias: {
				aliasA: "a",
				b$: `a${obps}index`,
				c$: `${absoluteOsBasedPath}a${obps}index`,
				multiAlias: ["b", "c", "d", "e", "a"],
				recursive: `recursive${obps}dir`,
				[`${absoluteOsBasedPath}d${obps}dir`]: `${absoluteOsBasedPath}c${obps}dir`,
				[`${absoluteOsBasedPath}d${obps}index.js`]: `${absoluteOsBasedPath}c${obps}index`,
				// alias configuration should work
				"#": `${absoluteOsBasedPath}c${obps}dir`,
				"@": `${absoluteOsBasedPath}c${obps}dir`,
				ignored: false
			},
			modules: absoluteOsBasedPath,
			useSyncFileSystemCalls: true,
			//@ts-ignore
			fileSystem: fileSystem
		});
	});

	it("should resolve a not aliased module", () => {
		expect(resolver.resolveSync({}, absoluteOsBasedPath, "a")).toEqual(
			`${absoluteOsBasedResolvedPath}a${posixSep}index`
		);
		expect(
			resolver.resolveSync({}, absoluteOsBasedPath, `a${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}index`);
		expect(
			resolver.resolveSync({}, absoluteOsBasedPath, `a${obps}dir`)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}dir${posixSep}index`);
		expect(
			resolver.resolveSync({}, absoluteOsBasedPath, `a${obps}dir${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}dir${posixSep}index`);
	});
	it("should resolve an aliased module", () => {
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, "aliasA")
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}index`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `aliasA${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}index`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `aliasA${obps}dir`)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}dir${posixSep}index`);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`aliasA${obps}dir${obps}index`
			)
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}dir${posixSep}index`);
	});
	it('should resolve "#" alias', () => {
		expect(resolver.resolveSync({}, `${absoluteOsBasedPath}`, "#")).toEqual(
			`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`
		);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `#${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`);
	});
	it('should resolve "@" alias', () => {
		expect(resolver.resolveSync({}, `${absoluteOsBasedPath}`, "@")).toEqual(
			`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`
		);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `@${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`);
	});
	it("should resolve an ignore module", () => {
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, "ignored")
		).toEqual(false);
	});
	it("should resolve a recursive aliased module", () => {
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, "recursive")
		).toEqual(
			`${absoluteOsBasedResolvedPath}recursive${posixSep}dir${posixSep}index`
		);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`recursive${obps}index`
			)
		).toEqual(
			`${absoluteOsBasedResolvedPath}recursive${posixSep}dir${posixSep}index`
		);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `recursive${obps}dir`)
		).toEqual(
			`${absoluteOsBasedResolvedPath}recursive${posixSep}dir${posixSep}index`
		);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`recursive${posixSep}dir${posixSep}index`
			)
		).toEqual(
			`${absoluteOsBasedResolvedPath}recursive${posixSep}dir${posixSep}index`
		);
	});
	it("should resolve a file aliased module", () => {
		expect(resolver.resolveSync({}, `${absoluteOsBasedPath}`, "b")).toEqual(
			`${absoluteOsBasedResolvedPath}a${posixSep}index`
		);
		expect(resolver.resolveSync({}, `${absoluteOsBasedPath}`, "c")).toEqual(
			`${absoluteOsBasedResolvedPath}a${posixSep}index`
		);
	});
	it("should resolve a file aliased module with a query", () => {
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, "b?query")
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}index?query`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, "c?query")
		).toEqual(`${absoluteOsBasedResolvedPath}a${posixSep}index?query`);
	});
	it("should resolve a path in a file aliased module", () => {
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `b${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}b${posixSep}index`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `b${obps}dir`)
		).toEqual(`${absoluteOsBasedResolvedPath}b${posixSep}dir${posixSep}index`);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`b${obps}dir${obps}index`
			)
		).toEqual(`${absoluteOsBasedResolvedPath}b${posixSep}dir${posixSep}index`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `c${obps}index`)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}index`);
		expect(
			resolver.resolveSync({}, `${absoluteOsBasedPath}`, `c${obps}dir`)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`c${obps}dir${obps}index`
			)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`);
	});
	it("should resolve a file aliased file", () => {
		expect(resolver.resolveSync({}, `${absoluteOsBasedPath}`, "d")).toEqual(
			`${absoluteOsBasedResolvedPath}c${posixSep}index`
		);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`d${obps}dir${obps}index`
			)
		).toEqual(`${absoluteOsBasedResolvedPath}c${posixSep}dir${posixSep}index`);
	});
	it("should resolve a file in multiple aliased dirs", () => {
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`multiAlias${obps}dir${obps}file`
			)
		).toEqual(`${absoluteOsBasedResolvedPath}e${posixSep}dir${posixSep}file`);
		expect(
			resolver.resolveSync(
				{},
				`${absoluteOsBasedPath}`,
				`multiAlias${obps}anotherDir`
			)
		).toEqual(
			`${absoluteOsBasedResolvedPath}e${posixSep}anotherDir${posixSep}index`
		);
	});
	it("should log the correct info", done => {
		const log = [];
		resolver.resolve(
			{},
			`${absoluteOsBasedPath}`,
			`aliasA${obps}dir`,
			{ log: v => log.push(v) },
			(err, result) => {
				if (err) return done(err);

				expect(result).toEqual(
					`${absoluteOsBasedResolvedPath}a${posixSep}dir${posixSep}index`
				);
				expect(log).toMatchSnapshot(path.sep === "/" ? "posix" : "win32");

				done();
			}
		);
	});

	it("should work with absolute paths", done => {
		const resolver = ResolverFactory.createResolver({
			alias: {
				[path.resolve(__dirname, "fixtures", "foo")]: false
			},
			modules: path.resolve(__dirname, "fixtures"),
			fileSystem: nodeFileSystem
		});

		resolver.resolve({}, __dirname, `foo${obps}index`, {}, (err, result) => {
			if (err) done(err);
			expect(result).toEqual(false);
			done();
		});
	});
});
