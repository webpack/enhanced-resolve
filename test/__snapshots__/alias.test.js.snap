// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alias should log the correct info 1`] = `
Array [
  "resolve 'aliasA/dir' in '/'",
  "  Parsed request is a module",
  "  No description file found in / or above",
  "  aliased with mapping 'aliasA': 'a' to 'a/dir'",
  "    Parsed request is a module",
  "    No description file found in / or above",
  "    resolve as module",
  "      looking for modules in /",
  "        existing directory /a",
  "          No description file found in /a or above",
  "          No description file found in /a or above",
  "          no extension",
  "            /a/dir is not a file",
  "          .js",
  "            /a/dir.js doesn't exist",
  "          .json",
  "            /a/dir.json doesn't exist",
  "          .node",
  "            /a/dir.node doesn't exist",
  "          as directory",
  "            existing directory /a/dir",
  "              No description file found in /a/dir or above",
  "              using path: /a/dir/index",
  "                No description file found in /a/dir or above",
  "                no extension",
  "                  existing file: /a/dir/index",
  "                    reporting result /a/dir/index",
]
`;

exports[`alias should log the correct info: posix 1`] = `
Array [
  "resolve 'aliasA/dir' in '/'",
  "  Parsed request is a module",
  "  No description file found in / or above",
  "  aliased with mapping 'aliasA': 'a' to 'a/dir'",
  "    Parsed request is a module",
  "    No description file found in / or above",
  "    resolve as module",
  "      looking for modules in /",
  "        existing directory /a",
  "          No description file found in /a or above",
  "          No description file found in /a or above",
  "          no extension",
  "            /a/dir is not a file",
  "          .js",
  "            /a/dir.js doesn't exist",
  "          .json",
  "            /a/dir.json doesn't exist",
  "          .node",
  "            /a/dir.node doesn't exist",
  "          as directory",
  "            existing directory /a/dir",
  "              No description file found in /a/dir or above",
  "              using path: /a/dir/index",
  "                No description file found in /a/dir or above",
  "                no extension",
  "                  existing file: /a/dir/index",
  "                    reporting result /a/dir/index",
]
`;

exports[`alias should log the correct info: win32 1`] = `
Array [
  "resolve 'aliasA/dir' in 'X:/'",
  "  Parsed request is a module",
  "  No description file found in X:/ or above",
  "  aliased with mapping 'aliasA': 'a' to 'a/dir'",
  "    Parsed request is a module",
  "    No description file found in X:/ or above",
  "    resolve as module",
  "      looking for modules in X:/",
  "        existing directory X:/a",
  "          No description file found in X:/a or above",
  "          No description file found in X:/a or above",
  "          no extension",
  "            X:/a/dir is not a file",
  "          .js",
  "            X:/a/dir.js doesn't exist",
  "          .json",
  "            X:/a/dir.json doesn't exist",
  "          .node",
  "            X:/a/dir.node doesn't exist",
  "          as directory",
  "            existing directory X:/a/dir",
  "              No description file found in X:/a/dir or above",
  "              using path: X:/a/dir/index",
  "                No description file found in X:/a/dir or above",
  "                no extension",
  "                  existing file: X:/a/dir/index",
  "                    reporting result X:/a/dir/index",
]
`;
